src/python/gem5/resources/resource.py

```diff
@@ -722,16 +725,24 @@ class SuiteResource(AbstractResource):
 
     def __init__(
         self,
-        workloads: Dict["WorkloadResource", Set[str]] = {},
+        workloads: Dict[Tuple[str, str], Set[str]] = {},
```

Comment:
Why is the key not a `WorkloadResource` object? 

Wouldn't `WorkloadResource` also have the same information (id and version)?

Reply:
I believe the thinking there is the entry for the Suite in the database only stores the resource_info in a simple JSON form. something like:

```json
workloads: { id: "hello-world", version: "v2.0.1"}
```

By having the signature of the `Suite`'s constructer mirror that of the databse enter you can pass the entry directly to the constructor.

At present we have some special code for suites to inspect the database entry `Workload` objects constructed the Workload objects, then pass it to the constructor. To me this makes logical sense but I can see why it's a pain to translate the database entry to the object.


My argument for keeping this as an object is as follows: This design assumes construction of a suite solely through the `obtain_resource` function. However, i think this should be allowed (and frankly, encouraged in some circumstances):

```py
workload_1 = Workload(...)
workload_2 = Workload(...)

my_suite = Suite({workload_1, workload_2})
```

For suites i'll also saw that allowing that above is conceptually much simpler than something like:

```py
workload_1 = Workload(id=1, version=2)
... # Some logic here to load `workload_1` into a suite

my_suite = Suite({id=1, version=2 ...})
```

I don't personally envision users being happy with a design that mandates they made an entry in a JSON file and load that JSON file and use `obtain_resources` to construct it.

What I see the `obtain_resource` function as doing (and the "clients" infastructure it depends on) is handle the translation between sources and the resource objects.

The trade off here is between having a constructor that represents our schema, or a constructor that an end user can use intuitively. The former has benefits. For example, it means if we update the schema we have less code that needs updated. However, I think the value in intuitive constructors that work in more scenarios is important.

Reply:
I changed it back so that we make all the workload resource objects so now the workloads_obj is Dict[WorkloadResource, Set[str]] like it used to be before. I am jsut getting all the workloads with a single call to the database and parsing that to create thins dictionary. This also only downloads the smaller resoruces(binaries, etc) only when required that is when `get_local_path` is called. 

The only thing to note with this solution is that I have the specific logic for parsing dependent workloads outside the workload class but I think this is better than passing a lot of unecessay variables in the SuitResource cinstructor.